src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -1125,6 +1139,77 @@ class FunctionBinaryArithmetic : public IFunction
         return function->execute(arguments, result_type, input_rows_count);
     }
 
+    ColumnPtr executeArrayImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const
+    {
+        const auto * return_type_array = checkAndGetDataType<DataTypeArray>(result_type.get());
+
+        if (!return_type_array)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Return type for function {} must be array.", getName());
+
+        if constexpr (!is_plus && !is_minus)
+            throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "Cannot use this operation on arrays");
```

Comment:
I think this should be checked in `getReturnTypeImplStatic`, because:
1. We have every information to do the check there
2. If we do the check here, that basically means the function successfully determines the return type of illegal operations (e.g.: `[1,2] / [3,4]`), but it cannot be executed